File: backend/app/schemas/compliance_scope.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


ScopeType = Literal[
    "ip_cidr", "ip_range", "mac_prefix_arp", "mac_prefix_ipguard",
    "ip_cidr_any", "ip_range_any", "mac_prefix_any",
]


class ComplianceScopeBase(BaseModel):
    scope_type: ScopeType = Field(
        ...,
        description=(
            "Scope type: ip_cidr, ip_range, mac_prefix_arp, mac_prefix_ipguard, "
            "ip_cidr_any, ip_range_any, or mac_prefix_any"
        )
    )
    scope_value: str = Field(
        ...,
        description="Scope value: e.g., 192.168.0.0/16, 192.168.1.1-255, AA:BB:CC"
    )
    description: str | None = Field(default=None, description="Description")


class ComplianceScopeCreate(ComplianceScopeBase):
    pass


class ComplianceScopeUpdate(BaseModel):
    scope_type: ScopeType | None = None
    scope_value: str | None = None
    description: str | None = None
    is_active: bool | None = None


class ComplianceScopeResponse(ComplianceScopeBase):
    id: int
    is_active: bool
    created_by: str
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class ComplianceScopeListResponse(BaseModel):
    items: list[ComplianceScopeResponse]
    total: int

File: frontend/src/api/complianceScope.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '@/lib/api';
import { API_ENDPOINTS } from '@/lib/constants';

export interface ComplianceScope {
  id: number;
  scope_type: 'ip_cidr' | 'ip_range' | 'mac_prefix';
  scope_value: string;
  description: string | null;
  is_active: boolean;
  created_by: string;
  created_at: string;
  updated_at: string | null;
}

export interface ComplianceScopeCreate {
  scope_type: 'ip_cidr' | 'ip_range' | 'mac_prefix';
  scope_value: string;
  description?: string;
}

export interface ComplianceScopeUpdate {
  scope_type?: 'ip_cidr' | 'ip_range' | 'mac_prefix';
  scope_value?: string;
  description?: string;
  is_active?: boolean;
}

export interface ComplianceScopeListResponse {
  items: ComplianceScope[];
  total: number;
}

async function fetchScopes(is_active?: boolean): Promise<ComplianceScopeListResponse> {
  const params = is_active !== undefined ? `?is_active=${is_active}` : '';
  const response = await apiClient.get(`${API_ENDPOINTS.COMPLIANCE_SCOPE}${params}`);
  return response.data;
}

export function useComplianceScopes(is_active?: boolean) {
  return useQuery({
    queryKey: ['compliance-scopes', is_active],
    queryFn: () => fetchScopes(is_active),
  });
}

export function useCreateComplianceScope() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (data: ComplianceScopeCreate) =>
      apiClient.post(API_ENDPOINTS.COMPLIANCE_SCOPE, data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['compliance-scopes'] });
    },
  });
}

export function useUpdateComplianceScope() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ id, data }: { id: number; data: ComplianceScopeUpdate }) =>
      apiClient.put(`${API_ENDPOINTS.COMPLIANCE_SCOPE}${id}`, data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['compliance-scopes'] });
    },
  });
}

export function useDeleteComplianceScope() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: number) =>
      apiClient.delete(`${API_ENDPOINTS.COMPLIANCE_SCOPE}${id}`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['compliance-scopes'] });
    },
  });
}

export function useToggleComplianceScope() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: number) =>
      apiClient.post(`${API_ENDPOINTS.COMPLIANCE_SCOPE}${id}/toggle`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['compliance-scopes'] });
    },
  });
}
